packages/svelte/src/internal/client/dom/blocks/await.js

```diff
@@ -66,41 +67,62 @@ export function await_block(anchor, get_input, pending_fn, then_fn, catch_fn) {
 		return e;
 	}
 
+	function pause() {
+		if (pending_effect) pause_effect(pending_effect);
+		if (then_effect) pause_effect(then_effect);
+		if (catch_effect) pause_effect(catch_effect);
+	}
+
 	const effect = block(() => {
 		if (input === (input = get_input())) return;
 
 		if (is_promise(input)) {
-			const promise = /** @type {Promise<any>} */ (input);
-
-			if (pending_fn) {
-				if (pending_effect && (pending_effect.f & INERT) === 0) {
-					destroy_effect(pending_effect);
-				}
+			let was_hydrating = hydrating;
 
+			if (hydrating && pending_fn) {
 				pending_effect = branch(() => pending_fn(anchor));
 			}
 
-			if (then_effect) pause_effect(then_effect);
-			if (catch_effect) pause_effect(catch_effect);
+			var promise = /** @type {Promise<any>} */ (input);
+
+			var settled = false;
 
 			promise.then(
 				(value) => {
+					settled = true;
 					if (promise !== input) return;
-					if (pending_effect) pause_effect(pending_effect);
+					pause();
 
 					if (then_fn) {
 						then_effect = create_effect(then_fn, value);
 					}
 				},
 				(error) => {
+					settled = true;
 					if (promise !== input) return;
-					if (pending_effect) pause_effect(pending_effect);
+					pause();
 
 					if (catch_fn) {
 						catch_effect = create_effect(catch_fn, error);
 					}
 				}
 			);
+
+			Promise.resolve().then(() => {
```

Comment:
This should use something that plays nicely with `flushSync` - `queue_micro_task` for example

Reply:
but does it matter in practise? Is this something worth considering? Who would have a resolved promise on mount? Even switching between resolved promises is an edge case, a resolved promise on mount even moreso.

Reply:
You could imagine a case where a SvelteKit `load` function does this...

```js
export function load() {
  return {
    critical: 'some critical data',
    streamed: get_promise()
  };
}
```

...and the promise resolves before hydration begins. Or a case where you have a client-side cache. I don't think these are the majority of cases, but I don't think they're edge cases either